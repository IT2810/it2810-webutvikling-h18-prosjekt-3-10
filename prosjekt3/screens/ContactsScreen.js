import React, { Component } from 'react';
import AddContactModal from '../components/AddContactModal';
import ContactInformationModal from '../components/ContactInformationModal';

import {
  Alert,
  ScrollView,
  View,
  AsyncStorage
} from 'react-native';

import { Contacts, Permissions } from 'expo';
import ContactsList from '../components/ContactsList';
import Loading from '../components/Loading';

export default class ContactsScreen extends Component {
  static navigationOptions = {
    title: "Contacts",
  };

  state = {
    contactModalVisible: false,
    addContactModalVisible: false,
    importedContacts: [],
    addedContacts: [],
    fetchingContacts: true,
  }

  componentDidMount = () => {
    this.loadContacts()
    // Check if user already has accepted sharing contacts
    // If not the contacts page will show an option to import from contacts
    // which will in turn ask for Contacts-permission. 
    Permissions.getAsync(Permissions.CONTACTS)
    .then(({status}) => {
      if (status === 'granted') {
        this.importContacts();
      }
    })
  }
  
  // Saves a contact in AsyncStorage
  saveContact = contact => {
    AsyncStorage.getItem('contacts')
    .then(result => {
      if (result == null) {
        return []
      } else {
        return JSON.parse(result)
      }
    })
    .then(data => {
      data.push(contact)
      data.sort((a,b) => a.firstName < b.firstName ? -1 : 1) 
      this.setState({
        addedContacts: data,
      })
      return JSON.stringify(data)
    })
    .then(contacts => {
      AsyncStorage.setItem('contacts', contacts)
    })
  }

  // Removes contact from AsyncStorage
  removeContact = contact => {
    // First gets all the saved contacts
    AsyncStorage.getItem('contacts')
    .then(result => {
      if (result == null) {
        console.error("There's no saved contacts")
        return []
      } else {
        return JSON.parse(result)
      }
    })
    // Then deletes the contact from that list
    .then(data => {
      console.log("Removing contact", contact)
      data = data.filter(savedContact => savedContact.id != contact.id)
      this.setState({
        addedContacts: data,
      })
      return JSON.stringify(data)
    })
    // And finally updates storage with the contact removed
    .then(contacts => {
      AsyncStorage.setItem('contacts', contacts)
    })
  }
  
  // Loads added contacts from AsyncStorage
  loadContacts = () => {
    let addedContacts = []
    AsyncStorage.getItem('contacts')
    .then(result => {
      if (result !== null) {
        addedContacts = JSON.parse(result)
      }
      this.setState({
        addedContacts,
        fetchingContacts: false,
      })
    })
  }

  // Imports contacts from phone contact list
  importContacts = async () => {
    const { status } = await Permissions.askAsync(Permissions.CONTACTS);
    if (status === 'granted') {
      result = await Contacts.getContactsAsync();
      importedContacts = 
        result.data
        // Filter away the contacts without any name
        .filter(contact => contact.firstName || contact.lastName)  
        // Give each contact-object an empty string for first or last name if they lack it
        .map(contact => {  
          const {firstName, lastName} = contact
          contact.firstName = firstName ? firstName : ""
          contact.lastName = lastName ? lastName : ""
          return contact
        })
        // Sort by first name
        .sort((a, b) => a.firstName < b.firstName ? -1 : 1) 
      this.setState({
        importedContacts,
        fetchingContacts: false,
      })
    } else {
      console.error('Contact read permission not granted, no changes made')
    }
  }

  handleDelete = (contact) => {
    Alert.alert(
      `Delete ${contact.firstName} ${contact.lastName}?`,
      "This will remove the contact",
      [
        {text: 'Cancel', onPress: () => {}, style: 'cancel'},
        {text: 'YES', onPress: () => {
          this.removeContact(contact)
        }},
      ],
      { cancelable: false }
    )
  }

  handleContactPress = contact => {
    this.setState({
      contactModalVisible: true,
      activeContact: contact,
    })
  }

  render() {
    return (
      <View>
        <ScrollView keyboardShouldPersistTaps='handled'>
          {
            // This checks if phone is currently fetching contacts and if so displays loading, if not displays list of contacts
            this.state.fetchingContacts 
            ?
            <Loading /> 
            :
            <ContactsList 
              addedContacts={this.state.addedContacts}
              importedContacts={this.state.importedContacts}
              handleContactPress={this.handleContactPress}
              handleDelete={this.handleDelete}
              importContacts={this.importContacts}
              addContact={() => this.setState({addContactModalVisible: true})}
            />
          }

            {/*Modals not shown on screen until their visibility is set to true in state*/}
            <ContactInformationModal contact={this.state.activeContact} closeCallback={() => this.setState({contactModalVisible: false})} visible={this.state.contactModalVisible} />
            <AddContactModal onSave={(contact) => this.saveContact(contact)} closeCallback={() => this.setState({addContactModalVisible: false})} visible={this.state.addContactModalVisible}/>
        </ScrollView>
      </View>
    );
  }
}