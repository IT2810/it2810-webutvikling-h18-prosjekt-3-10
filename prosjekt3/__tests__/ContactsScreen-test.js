import 'react-native';
import React from 'react';
import ContactsScreen from '../screens/ContactsScreen';
import renderer from 'react-test-renderer';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';

import MockAsyncStorage from 'mock-async-storage';

const mock = () => {
    const mockImpl = new MockAsyncStorage()
    jest.mock('AsyncStorage', () => mockImpl)
}
const release = () => jest.unmock('AsyncStorage')

import { AsyncStorage } from 'react-native';

describe('ContactsScreen snapshot', () => {
  jest.useFakeTimers();
  beforeEach(() => {
    NavigationTestUtils.resetInternalState();
  });

  it('renders the contacts screen', async () => {
    const tree = renderer.create(<ContactsScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Unit testing', () => {
    let contactsScreen = renderer.create(<ContactsScreen />).getInstance();
    const contact = {
        firstName: "Hello",
        lastName: "World",
        id: "1231231231",
        number: "02020202",
    }
    it('should save contact in state', async () => {
        mock();
        await contactsScreen.saveContact(contact);
        expect(contactsScreen.state.addedContacts.length).toEqual(1);
        expect(contactsScreen.state.addedContacts[0].firstName).toEqual("Hello")
    })
    it('should have saved contact in AsyncStorage', async () => {
        const value = await AsyncStorage.getItem('contacts')
        expect(value).toBe(JSON.stringify([contact]))
    })
    it('should allow several contacts in AsyncStorage', async () => {
        await contactsScreen.saveContact(contact);
        const value = await AsyncStorage.getItem('contacts')
        expect(value).toBe(JSON.stringify([contact, contact]))
    })
    it('should be able to load contacts from AsyncStorage', async () => {
        contactsScreen.setState({addedContacts: []})
        expect(contactsScreen.state.addedContacts.length).toEqual(0)        
        await contactsScreen.loadContacts();
        expect(contactsScreen.state.addedContacts.length).toBeGreaterThan(0)
    }) 
    it('should remove contacts from AsyncStorage', async () => {
        // This should delete all the contacts as they all share the same id
        await contactsScreen.removeContact(contact)
        const value = await AsyncStorage.getItem('contacts')
        expect(value).toBe("[]")
    })
    it('should handle contact press', () => {
        contactsScreen.handleContactPress(contact);
        expect(contactsScreen.state.contactModalVisible).toBe(true)
        expect(contactsScreen.state.activeContact).toEqual(contact)
    })
});

release();
