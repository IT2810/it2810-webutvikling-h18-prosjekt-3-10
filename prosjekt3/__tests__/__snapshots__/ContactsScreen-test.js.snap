// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsScreen snapshot renders the contacts screen 1`] = `<View />`;
